src/aeso_stdlib.erl

```diff
@@ -0,0 +1,199 @@
+".
+
+std_list() ->
+"
+namespace List =
+
+  function empty(l) = switch(l)
+    [] => true
+    _  => false
+
+
+  function first(l) = switch(l)
+    [] => None
+    h::_ => Some(h)
+
+  function tail(l) = switch(l)
+    [] => None
+    _::t => Some(t)
+
+  function last(l) = switch(l)
+    [] => None
+    [x] => Some(x)
+    _::t => last(t)
+
+  function find(p, l) = switch(l)
+    [] => None
+    h::t => if(p(h)) Some(h) else find(p, t)
+
+  function nth(n, l) = switch(l)
+    h::t => if(n == 0) Some(h) else nth(n-1, t)
+    [] => None
+
+  function get(n, l) = switch(l)
+    [] => abort(\"Out of index get\")
+    h::t => if(n == 0) h else get(n-1, t)
+
+
+  function length(l) = length_(l, 0)
+  private function length_(l, acc) = switch(l)
+    [] => acc
+    _::t => length_(t, acc + 1)
+
+  function foldr(cons, nil, l) = switch(l)
+    [] => nil
+    h::t => cons(h, foldr(cons, nil, t))
+
+  function foldl(rcons, acc, l) = switch(l)
+    [] => acc
+    h::t => foldl(rcons, rcons(acc, h), t)
+
+
+  function reverse(l) = foldl((lst, el) => el :: lst, [], l)
+
+
+  function map(f, l) = map_(f, l, [])
+  private function map_(f, l, acc) = switch(l)
+    [] => reverse(acc)
+    h::t => map_(f, t, f(h)::acc)
+
+  function flat_map(f, l) = switch(l)
+    [] => []
+    h::t => f(h) ++ flat_map(f, t)
+
+
+  function filter(p, l) = filter_(p, l, [])
+  private function filter_(p, l, acc) = switch(l)
+    [] => reverse(acc)
+    h::t => filter_(p, t, if(p(h)) h::acc else acc)
+
+  /* Take `n` first elements */
+  function take(n, l) = if(n < 0) abort(\"Take negative number of elements\") else take_(n, l, [])
+  private function take_(n, l, acc) =
+    if(n == 0) reverse(acc)
+    else switch(l)
+      [] => reverse(acc)
+      h::t => take_(n-1, t, h::acc)
+
+  /* Drop `n` first elements */
+  function drop(n, l) =
+   if(n < 0) abort(\"Drop negative number of elements\")
+   elif (n == 0) l
+   else switch(l)
+      [] => []
+      h::t => drop(n-1, t)
+
+  /* Get the longest prefix of a list in which every element matches predicate `p` */
+  function take_while(p, l) = take_while_(p, l, [])
+  private function take_while_(p, l, acc) = switch(l)
+    [] => reverse(acc)
+    h::t => if(p(h)) take_while_(p, t, h::acc) else reverse(acc)
+
+  /* Drop elements from `l` until `p` holds */
+  function drop_while(p, l) = switch(l)
+    [] => []
+    h::t => if(p(h)) drop_while(p, t) else l
+
+  /* Splits list into two lists of elements that respectively match and don't match predicate `p` */
+  function partition(p, l) = partition_(p, l, [], [])
+  private function partition_(p, l, acc_t, acc_f) = switch(l)
+    [] => (reverse(acc_t), reverse(acc_f))
+    h::t => if(p(h)) partition_(p, t, h::acc_t, acc_f) else partition_(p, t, acc_t, h::acc_f)
+
+
+  function concats(ll) = foldr((l1, l2) => l1 ++ l2, [], ll)
+
+  function all(p, l) = foldl((prev, next) => prev && p(next), true, l)
+
+  function any(p, l) = foldl((prev, next) => prev || p(next), false, l)
```

Comment:
We probably want short-cut behaviour for these.

Reply:
What do you mean?

Reply:
That `all` only evaluates `p` until it gets a `false`, in the same way that `false && x` doesn't evaluate `x`.

Reply:
fixed